### PR TITLE
fix(agent/publishAsync): seal on-chain publish via publisher-fallback flag

### PR DESCRIPTION
**Pull request by zsculac**
## Summary

PR #436 made `precomputedAttestation` mandatory for V10 on-chain publish. The sync `_publish` path was updated to inline-mint the seal. The async `publishAsync` path — used by `POST /api/epcis/capture` via `AsyncPublisher` — was not, so every EPCIS capture queued a lift that the publisher rejected with `"publish requires precomputedAttestation"` and downgraded to terminal-failed.

Reproduction: `pnpm test:devnet:agent-provenance` is **unaffected** because it goes through `_publish`. The EPCIS demo regression surfaces against the v10.0.0-rc.6 daemon:

```
$ API_PORT_BASE=9301 DEVNET_ENABLE_PUBLISHER=1 DEVNET_NO_AUTH=1 ./scripts/devnet.sh start 5
$ cd demo/epcis-bike && node run.mjs --no-pause
Aggregate — Finalized: 0 · Failed: 7 · Still pending: 0   ← regression
```

## Fix

```
   agent.publishAsync(...)
        │  sets allowPublisherFallbackSeal: true on LiftRequest
        │  when chain is V10-capable + CG on-chain
        ▼
   asyncPublisher.lift(LiftRequest)
        │  persists flag through job.request via existing JSON-stringify path
        ▼
   publishExecutor → DKGPublisher.publish(PublishOptions)
        │  if no precomputedAttestation + allowPublisherFallbackSeal + signer:
        │    mint EIP-712 over (chainId, kav10Address, cgId, kcMerkleRoot, publisherAddress)
        │    sign via publisherFallbackAuthorAddress() + signAuthorAttestationAsPublisher()
        │  else if no precomputedAttestation: existing fall-through to tentative
        │  else: existing preflight validation
        ▼
   chain
```

Six files, +182 / −1.

| File | Change |
|---|---|
| `packages/publisher/src/lift-job-types.ts` | New `LiftRequest.allowPublisherFallbackSeal?: boolean` |
| `packages/publisher/src/publisher.ts` | New `PublishOptions.allowPublisherFallbackSeal?: boolean` |
| `packages/publisher/src/async-lift-publish-options.ts` | Thread flag through `mapLiftRequestToPublishOptions` |
| `packages/publisher/src/dkg-publisher.ts` | After `kcMerkleRoot` compute, inline-mint fallback seal if flagged |
| `packages/agent/src/dkg-agent.ts` | `publishAsync` sets flag when V10 chain + CG on-chain |
| `packages/agent/test/publish-jsonld.test.ts` | Two new tracer tests against real Hardhat |

## Why publisher-side instead of agent-side

First attempt was to mirror `_publish` and mint at the agent layer (3 rounds: graph-stripping per `selectQuadsForRoots`, root-ordering per `parseStoredPublicQuads`, post-partition shape per `partitionPublishAsyncQuads`). Merkles kept diverging from the publisher's recompute. The async workspace round-trip (`selectQuadsForRoots` strips `graph: ''`, `parseStoredPublicQuads` reassembles per-root through JSON literals in `request.roots` order) makes byte-perfect prediction at agent-side brittle.

Centralising the merkle compute on the publisher side — where it already lives for the sealed-path — eliminates the drift class entirely. **Same input → same merkle by construction**, so the existing seal-integrity guard at `dkg-publisher.ts:1898` cannot trip for the fallback path.

**Caller-supplied seals still take priority** and are still validated — the strict mode PR #436 established for `/api/shared-memory/publish` is preserved.

## Author attribution

KCs published via this path are attributed to the publisher's own EOA — RFC-001 §4 mode (a), *"daemon signs as itself"*. Per-event author attribution for `/api/epcis/capture` (bearer-token agent / `authorAgentAddress` body field / `preSignedAuthorAttestation`) is **deferred to a follow-up**. Request body shape stays unchanged in this PR.

## Tests

- `packages/agent/test/publish-jsonld.test.ts` — two new vitest cases against a real Hardhat fixture: one for envelope `{public}`, one for bare (private-only) JSON-LD that mirrors the EPCIS capture shape. Both assert `job.request.allowPublisherFallbackSeal === true`.
- Full `publish-jsonld.test.ts` suite green (12/12).
- `async-lift-publisher.integration.test.ts`, `async-lift-publish-options.test.ts`, `async-lift-runner.test.ts` green (24/24).
- `agent-provenance-e2e.test.ts` green (15/15, 1 skipped pre-existing).
- **5-node devnet end-to-end:** `API_PORT_BASE=9301 DEVNET_ENABLE_PUBLISHER=1 DEVNET_NO_AUTH=1 ./scripts/devnet.sh start 5` then `cd demo/epcis-bike && node run.mjs --no-pause` reports `Aggregate — Finalized: 7 · Failed: 0 · Still pending: 0`, Phase 6 grant succeeds, Phase 7 cross-node anchor visibility passes.

## Pre-existing issue surfaced but NOT addressed here

The EPCIS query layer's owner-view composite (`dkg epcis query --finalized=false` and the default finalized variant) returns `0 event(s)` even though raw SPARQL against the canonical partition shows the anchors landed. Reproduces on `main` as well as this branch, so **not introduced by this fix** — tracking as a separate ticket.

## Test plan

- [x] In-process Hardhat tracer tests (`pnpm --filter @origintrail-official/dkg-agent test`)
- [x] Publisher async-lift regression suite green
- [x] `agent-provenance-e2e.test.ts` green (covers sync `_publish` path — confirms no regression on the sealed surface)
- [x] 5-node devnet `demo/epcis-bike/run.mjs` end-to-end: `Finalized: 7 / Failed: 0`
- [ ] CI: full `pnpm test` suite on this PR
- [ ] Reviewer: confirm the fallback path preserves PR #436's strict-seal validation for caller-supplied seals (`/api/shared-memory/publish`, `/api/assertion/:name/write` → `publish`)
- [ ] Reviewer: confirm the publisher-fallback EOA is the desired author for unattributed EPCIS captures (vs. requiring explicit attribution via a future `authorAgentAddress` body extension)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/publisher/src/dkg-publisher.ts [hunk 1]
```diff
@@ -1857,6 +1859,47 @@ export class DKGPublisher implements Publisher {
...
+      // so the preflight below cannot mismatch.
+      if (
+        !options.precomputedAttestation
+        && options.allowPublisherFallbackSeal
+        && v10ChainId !== undefined
+        && v10KavAddress !== undefined
+        && publisherSigner !== undefined
+      ) {
+        try {
+          const fallbackAuthorAddress = await this.publisherFallbackAuthorAddress();
```
[the remark this answers is not in the document]
Comment (zsculac) [reply packages/publisher/src/dkg-publisher.ts]: Fixed in 92d8a386. Replaced the `publisherFallbackAuthorAddress()` + `signAuthorAttestationAsPublisher()` round-trip with direct use of `publisherSigner.address` + `publisherSigner.signTypedData()` — the signer is already cg-resolved at the publish call boundary, so `KC.author == tx submitter` holds regardless of `publisherAddressResolver` config.

Added `Diagram 12` tests in `agent-provenance-e2e.test.ts` that assert `getLatestMerkleRootAuthor(kcId).toLowerCase() === signingWallet.address.toLowerCase()` on a confirmed fallback publish.

File: packages/agent/test/publish-jsonld.test.ts [hunk 2]
```diff
@@ -284,6 +284,52 @@ describe('publishJsonLd', () => {
...
+          '@type': 'Thing',
+          'name': 'Async Seal Public',
+        },
+      },
+      { localOnly: true },
+    );
+
+    const asyncPublisher = new TripleStoreAsyncLiftPublisher(store);
+    const job = await asyncPublisher.getStatus(captureID);
+    expect(job?.request.allowPublisherFallbackSeal).toBe(true);
```
[the remark this answers is not in the document]
Comment (zsculac) [reply packages/agent/test/publish-jsonld.test.ts]: Fixed in 92d8a386. Added two end-to-end tests in `packages/publisher/test/agent-provenance-e2e.test.ts` (Diagram 12) that exercise the full `publish({allowPublisherFallbackSeal:true})` path against a real Hardhat deployment:

1. **`confirms and KC.author == publisherSigner.address`** — runs the publish flow end-to-end, asserts `result.status === 'confirmed'`, `kcId > 0n`, and reads back `getLatestMerkleRootAuthor(kcId)` to verify the publisher-side fallback mint produced a valid attestation.
2. **`is a no-op when a caller seal is also supplied`** — confirms PR #436's strict-seal preflight for caller-supplied seals (`/api/shared-memory/publish`) is preserved.

Regression in the mapper or the mint branch now fails at this level instead of slipping through to the 5-node devnet check.

File: packages/agent/src/dkg-agent.ts [hunk 3]
```diff
@@ -3754,6 +3754,14 @@ export class DKGAgent {
       }
     }
 
+    // Signal publisher-fallback sealing on V10 chains; publisher mints
+    // the seal at processNext-time. Author == publisher's own EOA.
+    const onChainId = await this.getContextGraphOnChainId(contextGraphId);
+    const allowPublisherFallback =
```
[the remark this answers is not in the document]
Comment (zsculac) [reply packages/agent/src/dkg-agent.ts]: Fixed in b6a3a0c0. Switched the gate from method-presence to `this.chain.isV10Ready?.() === true` (the canonical live probe also used by publisher-runner.ts). `NoChainAdapter.isV10Ready()` returns false even though the V10 methods exist as throwing stubs, so the persisted-job mis-attribution scenario no longer applies.

Added `async publish does NOT signal allowPublisherFallbackSeal when chain reports !isV10Ready` in `publish-jsonld.test.ts` that constructs an agent on a real Hardhat fixture, post-monkey-patches `isV10Ready` to return false, and asserts the flag is not set. RED before the fix, GREEN after.

File: packages/publisher/test/agent-provenance-e2e.test.ts [hunk 4]
```diff
@@ -821,3 +821,63 @@ describe('Diagram 11 — Phase 5 precomputedAttestation (sign-at-creation)', ()
...
+    const kcId = result.onChainResult!.batchId;
+    const onChainAuthor: string = await kcs().getLatestMerkleRootAuthor(kcId);
+    expect(onChainAuthor.toLowerCase()).toBe(author.address.toLowerCase());
+  });
+
+  it('publish({allowPublisherFallbackSeal:true}) is a no-op when a caller seal is also supplied', async () => {
+    // Caller-supplied seals still take priority. The fallback flag is
+    // ignored when the caller already minted; existing strict
+    // validation path runs unchanged.
+    const author = new ethers.Wallet(HARDHAT_KEYS.CORE_OP);
```
[the remark this answers is not in the document]
Comment (zsculac) [reply packages/publisher/test/agent-provenance-e2e.test.ts]: Fixed in b6a3a0c0. Switched `author` to `FIXTURE_AUTHOR_KEY` (a distinct fixture wallet, not `HARDHAT_KEYS.CORE_OP` used by `makePublisher()`) and added `expect(onChainAuthor).not.toBe(publisherAddress)` so the assertion now actually distinguishes "caller seal honoured" from "fallback overwrote it".

File: packages/publisher/src/lift-job-types.ts [hunk 5]
```diff
@@ -46,6 +48,7 @@ export const LIFT_REQUEST_IMMUTABLE_FIELDS = [
   'subGraphName',
   'accessPolicy',
   'allowedPeers',
+  'allowPublisherFallbackSeal',
```
[the remark this answers is not in the document]
Comment (zsculac) [reply packages/publisher/src/lift-job-types.ts]: Fixed in ca3eed7e. Updated the `LIFT_REQUEST_IMMUTABLE_FIELDS` expectation in `packages/publisher/test/lift-job-types.test.ts` to include `allowPublisherFallbackSeal`. Test now 8/8 GREEN.

File: packages/publisher/src/async-lift-publish-options.ts [hunk 6]
```diff
@@ -121,6 +121,7 @@ export function mapLiftRequestToPublishOptions(input: LiftPublishMappingInput):
     onPhase: input.resolved.onPhase,
     receiverSignatureProvider: input.resolved.receiverSignatureProvider,
     publishContextGraphId: input.resolved.publishContextGraphId,
+    allowPublisherFallbackSeal: input.request.allowPublisherFallbackSeal,
```
[the remark this answers is not in the document]
Comment (zsculac) [reply packages/publisher/src/async-lift-publish-options.ts]: Fixed in ca3eed7e. Plumbed the flag through both enqueue surfaces:

- `packages/cli/src/daemon/routes/publisher.ts` — `POST /api/publisher/enqueue` now reads `parsed.allowPublisherFallbackSeal` and threads it into the `publisherControl.lift({...})` reconstruction (coerced via `Boolean(...)` when present).
- `packages/cli/src/cli.ts` + `packages/cli/src/api-client.ts` — added optional `--allow-publisher-fallback-seal` CLI flag and the matching field on `ApiClient.publisherEnqueue`. Included in `enqueueFields` so both the daemon-API path and the in-process `inspector.publisher.lift({...enqueueFields, ...})` fallback carry it.

New TDD-style test `packages/cli/test/publisher-enqueue-route.test.ts` pins the wire contract: `true` is forwarded, `false` is forwarded, and the field is absent when the caller omits it. RED before the route fix, GREEN (3/3) after.

File: packages/agent/src/dkg-agent.ts [hunk 7]
```diff
@@ -3754,6 +3754,17 @@ export class DKGAgent {
...
 
+    // Signal publisher-fallback sealing on V10 chains; publisher mints
+    // the seal at processNext-time. Author == publisher's own EOA.
+    // Use `isV10Ready()` rather than method-presence — `NoChainAdapter`
+    // and partially-configured adapters expose the V10 methods as
+    // throwing stubs but report `isV10Ready() === false`, so a
+    // presence gate would persist the flag on a node that cannot
+    // mint a V10 seal.
+    const onChainId = await this.getContextGraphOnChainId(contextGraphId);
+    const allowPublisherFallback =
```
[the remark this answers is not in the document]
Comment (zsculac) [reply packages/agent/src/dkg-agent.ts]: Fixed in fe3b2c85. Pivoted to deferred-eval semantics: `agent.publishAsync` now always sets `allowPublisherFallbackSeal: true` on the LiftRequest. The publisher already re-resolves V10 readiness at `processNext()` time (`v10ChainId`, `v10KavAddress`, `publisherSigner` all derived inline) before minting — on non-V10 chains the fallback simply doesn't fire (Diagram 12 e2e covers this). Snapshotting was both redundant with the publisher gate and broken in exactly the persisted-job edge case you flagged.

Repurposed the round-2 `!isV10Ready` test in `packages/agent/test/publish-jsonld.test.ts` as `'async publish always signals allowPublisherFallbackSeal — V10 readiness is checked at processNext-time'`. Same monkey-patch, inverted assertion (flag is now `true` regardless of `isV10Ready`). RED before the agent change, GREEN after.

File: packages/cli/src/daemon/routes/publisher.ts [hunk 8]
```diff
@@ -405,6 +406,9 @@ export async function handlePublisherRoutes(ctx: RequestContext): Promise<void>
       transitionType,
       authority: { type: authorityType, proofRef },
       ...(priorVersion ? { priorVersion } : {}),
+      ...(allowPublisherFallbackSeal !== undefined
+        ? { allowPublisherFallbackSeal: Boolean(allowPublisherFallbackSeal) }
```
[the remark this answers is not in the document]
Comment (zsculac) [reply packages/cli/src/daemon/routes/publisher.ts]: Fixed in fe3b2c85. Replaced `Boolean(parsed.allowPublisherFallbackSeal)` with a strict `typeof === 'boolean'` check; non-boolean payloads (`"true"`, `"false"`, `1`, etc.) are treated as unset rather than coerced.

Added 3 coercion tests to `packages/cli/test/publisher-enqueue-route.test.ts` — `"true"`, `"false"`, and `1` must NOT appear on the captured LiftRequest. RED before the strict check (`Boolean("false")` is `true`), GREEN after. Total 6/6 ✓.

File: packages/publisher/src/dkg-publisher.ts [hunk 9]
```diff
@@ -1857,6 +1859,51 @@ export class DKGPublisher implements Publisher {
...
+            ...options,
+            precomputedAttestation: {
+              expectedMerkleRoot: kcMerkleRoot,
+              authorAddress: publisherSigner.address,
+              signature: { r: ethers.getBytes(sig.r), vs: ethers.getBytes(sig.yParityAndS) },
+              schemeVersion: AUTHOR_SCHEME_VERSION_V1,
+            },
+          };
+          this.log.info(ctx, `Minted publisher-fallback AuthorAttestation (author=${publisherSigner.address})`);
+        } catch (err) {
```
[the remark this answers is not in the document]
Comment (zsculac) [reply packages/publisher/src/dkg-publisher.ts]: Fixed in fe3b2c85. Removed the try/catch around the fallback mint block; signing errors now propagate as hard rejections. The caller explicitly asked for fallback by setting `allowPublisherFallbackSeal: true`, so a signing failure here IS the publish failure — not a retryable transport problem.

Added `'publish({allowPublisherFallbackSeal:true}) fails fast when signTypedData throws'` to Diagram 12 in `packages/publisher/test/agent-provenance-e2e.test.ts`. Monkey-patches `getPublisherSigner` to wrap the real signer with a `signTypedData` that throws `'publisher signer offline'`; asserts `publisher.publish({...allowPublisherFallbackSeal:true}).rejects.toThrow(/publisher signer offline/)`. RED before (publish silently confirmed via the no-seal path on this Hardhat fixture), GREEN after.

**Review by lupuszr — APPROVED** [review]
